core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -556,17 +556,18 @@ private[spark] class Master(
     // in the queue, then the second app, etc.
     if (spreadOutApps) {
       // Try to spread out each app among all the nodes, until it has all its cores
-      for (app <- waitingApps if app.coresLeft > 0) {
+      for (app <- waitingApps if app.coresLeft >= app.desc.coreNumPerTask) {
+        val coreNumPerTask = app.desc.coreNumPerTask
         val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
           .filter(canUse(app, _)).sortBy(_.coresFree).reverse
         val numUsable = usableWorkers.length
         val assigned = new Array[Int](numUsable) // Number of cores to give on each node
         var toAssign = math.min(app.coresLeft, usableWorkers.map(_.coresFree).sum)
         var pos = 0
-        while (toAssign > 0) {
-          if (usableWorkers(pos).coresFree - assigned(pos) > 0) {
-            toAssign -= 1
-            assigned(pos) += 1
+        while (toAssign >= coreNumPerTask) {
```

Comment:
We should add a comment here:
```
// Take into account the number of cores the application uses per task
// to avoid starving executors or wasting cluster resources (SPARK-5337)
```